\documentclass[rmp,superscriptaddress,11pt]{revtex4-1}
\usepackage[utf8x]{inputenc}
\usepackage{amsmath,amsthm,amsfonts,amssymb,amscd}
\usepackage{graphicx}
\usepackage{wrapfig}
\usepackage{enumerate}
\usepackage{color}
\usepackage[final,
            colorlinks = true,
            linkcolor = blue,
            urlcolor  = blue,
            citecolor = blue]{hyperref}
\usepackage{fontspec}
\usepackage{natbib}
\bibliographystyle{plainnat}

\newcommand{\bla}[1]{\textit{bla}$_\mathrm{#1}$}

%\setmainfont{Helvetica}
\usepackage{unicode-math}
\setmathfont{XITS Math}


%\setlength{\parindent}{20pt}
\def\thesubsectiondis{\unskip\arabic{subsection}}
 
\begin{document}

\title{Visualizing and quantifying structural diversity around mobile AMR genes}
\author{Liam P. Shaw}
\thanks{Correspondence: liam.shaw@biology.ox.ac.uk}
\affiliation{Department of Biology, University of Oxford, Oxford, UK\looseness=-5}
\affiliation{Department of Biosciences, University of Durham, Durham, UK\looseness=-5}
\author{Richard A. Neher}
\affiliation{Biozentrum, University of Basel, Basel, Switzerland\looseness=-5}

\begin{abstract}
Understanding the evolution of mobile genes is important for understanding the spread of antimicrobial resistance (AMR). Many clinically important AMR genes have been mobilized by mobile genetic elements (MGEs) on the kilobase scale, such as integrons and transposons, which can integrate into both chromosomes and plasmids and lead to rapid spread of the gene through bacterial populations. Looking at the flanking regions of these mobile genes in diverse genomes can highlight common structures and reveal patterns of MGE spread. However, historically this has been a largely descriptive process, relying on gene annotation and expert knowledge. Here we describe a general method to visualize and quantify the structural diversity around genes using \textit{pangraph} to find blocks of homologous sequence. We apply this method to a set of twelve clinically important beta-lactamase genes and provide interactive visualizations of their flanking regions at \href{https://liampshaw.github.io/flanking-regions}{https://liampshaw.github.io/flanking-regions}. We show that nucleotide-level variation in the mobile gene itself generally correlates with increased structural diversity in its flanking regions, demonstrating a relationship between rates of mutational evolution and rates of structural evolution, and find a bias for greater structural diversity upstream. Our framework is a starting point to investigate general rules that apply to the horizontal spread of new genes through bacterial populations. 
\end{abstract}


\maketitle

\section{Introduction}

\noindent Mobile genetic elements (MGEs) allow the horizontal movement of genes within and between bacterial species, contributing to a vast range of bacterial phenotypes including antimicrobial resistance (AMR). A `mobile gene' will be found in multiple genomic contexts, where its flanking regions contain information on the evolutionary history of the MGE (or MGEs) that have mobilized it.\footnote{In line with common usage, by `mobile gene' we refer to any gene that can be or has been found on MGEs in the recent past, on the scale of decades, allowing for the possibility that it may not be \textit{currently} mobile in all genomes.} However, despite the increasing availability of complete genomes, high structural diversity in these flanking regions generated by both transposition and recombination means that analysing them is challenging. \par

Each mobile AMR gene has its own unique epidemiological history. As a recent example, the metallo-beta-lactamase gene \bla{NDM-1} was first reported by \textcite{Yong2009}. The earliest NDM-positive isolate dates only from 2005 \cite{Jones2014}, but already by the 2010s \bla{NDM-1} had been seen worldwide in diverse bacteria. As of 2023 there are public genomes containing \bla{NDM} genes from seventeen bacterial genera \cite{Alcock2023}. Such a rapid horizontal and global spread with multiple rearrangements presents a challenge for genomic epidemiology. The flanking regions around \bla{NDM-1} show large structural diversity, particularly upstream of the gene, although with some traces of a common ancestral MGE. \textcite{Acman2022} developed a methodology for iteratively `splitting' flanking sequences, finding that downstream patterns of structural diversity in a global dataset supported the previous conclusion that \bla{NDM-1} had been first mobilized by a Tn\textit{125} transposon \cite{Toleman2012}. This example demonstrates that analysis of flanking regions is a valuable but difficult task, often requiring bespoke methods. For this reason, almost all the existing literature on the flanking regions of mobile genes remains descriptive and focused on single genes at a time, meaning it is difficult to extrapolate the general rules {--} if any {--} that govern evolution in these regions. \par

The high levels of structural diversity around mobile genes present two challenges: visualization and quantification. Visualizations are often based on annotated gene clusters \cite{Gilchrist2021}, but genes are frequently disrupted in flanking regions. Other tools aim to identify discrete clusters for tracking MGE epidemiology. For example, TETyper was developed specifically for transposable MGEs and can identify small-scale changes associated with transposition \cite{Sheppard2018} and Flanker performs alignment-free clustering of flanking sequences based on mash distances \cite{Matlock2021}. Such tools are useful but difficult to connect to the processes that generate structural diversity. In general, our understanding of these processes remains qualitative.\par

Here, we aim to provide a starting point for quantitative analysis of structural diversity around mobile genes. We outline an annotation-free approach based on finding homologous sequence blocks with \textit{pangraph} \cite{Noll2023} that scales to thousands of sequences and connects visualization with quantification. As a demonstration, we apply our method to the flanking regions of twelve beta-lactamase genes.

\begin{figure*}[ht]
    \centering
    \includegraphics[width=\linewidth]{figs/figure-workflow.pdf}
    \caption{\textbf{Overview of pipeline and outputs.} (a) Schematic worfklow. Given a focal gene and a set of contigs as input, the pipeline extracts the gene's flanking regions (default: 5kb upstream/downstream) and pipeline uses \textit{pangraph} to build a pangenome graph. Optionally, annotations (in gff format) can also be included for visualisation. The output is then used for (b) interactive html visualizations of homology blocks for exploration and (c) statistics such as the upstream/downstream breakpoint distances between pairs of sequences (shown: inverse cumulative distribution function for pairwise comparisons) and the block diversity. }
    \label{fig:workflow}
\end{figure*}

\begin{table*}
\begin{small}
\begin{tabular}{p{0.15\linewidth}p{0.08\linewidth}p{0.28\linewidth}p{0.13\linewidth}p{0.17\linewidth}p{0.1\linewidth}}
 \textbf{Beta-lactamase} & \textbf{Group} & \textbf{First description} & \textbf{First genome} & \textbf{N (chrom/plas)} & \textbf{Genera}\\\hline
 CMY-2 & 1, 1e & \textcite{pmid2079378} & 2002 & 196 (47/149) & 7 \\
 CTX-M-15 & 2be & \textcite{pmid11470367}& 2001 & 746 (214/532) & 13\\
 CTX-M-65 & 2be & \textcite{pmid19073302} & 2008 & 434 (90/344) & 10 \\
 GES-1 & 2f & \textcite{pmid10681329} & 2008 & 27 (7/20) & 8 \\
 IMP-4 & 3a &  \textcite{pmid8141584}& 1998 & 33 (0/33) & 7 \\
 KPC-2 & 2f &  \textcite{pmid11257029} & 2007 & 515 (17/498) & 11 \\
 NDM-1 & 3a & \textcite{pmid19770275} & 2009 & 338 (56/282) & 14 \\
 OXA-10 & 2d & \textcite{pmid3126705} & 2001 & 93 (24/69) & 12 \\
 OXA-48 & 2df & \textcite{pmid14693513} & 2010 & 158 (66/92) & 4 \\
 PER-1 & 2be & \textcite{pmid8517722} & 2006 & 24 (16/8) & 7 \\
 TEM-1 & 2b & \textcite{pmid5326330} & 1974 & 1581 (362/1212) & 24 \\
 VIM-1 & 3a & \textcite{pmid10390207}  & 2001 & 80 (5/75) & 7\\
\end{tabular}
\end{small}
\caption{\textbf{Beta-lactamase genes used as focal genes for example analysis.} Complete contigs (chromosomes or plasmids) were taken from CARD prevalence database v3.1.0 and re-analysed to confirm gene presence. Only contigs with sufficient flanking sequence (+/- 5kb either side of the focal gene) with <25 SNVs in the focal gene. Functional group information is as in \textcite{Bush2010}. First descriptions are given for the specific named focal gene rather than the enzyme family as a whole - for example, the CTX-M family was first described in 1990 \cite{pmid2079378}.}
\label{table:bl-genes}
\end{table*}

\pagebreak 

\section{Methodology}


\subsection*{Overview of the pipeline}

\noindent Figure \ref{fig:workflow} gives a schematic overview of the pipeline available at \href{https://github.com/liampshaw/mobile-gene-regions}{https://github.com/liampshaw/mobile-gene-regions}. The use case is where an investigator has a \textit{focal gene} and a set of \textit{contigs} which putatively contain that gene (or closely-related variants). For example, the focal gene might be \bla{NDM-1} and the contigs would be assemblies containing \bla{NDM} gene variants downloaded from NCBI MicroBIGG-E with the search query \href{https://www.ncbi.nlm.nih.gov/pathogens/microbigge/#element_symbol:blaNDM*}{\texttt{element\_symbol:blaNDM*}}. The first step is to extract the flanking regions of the focal gene for some specified distance value (default: +/- 5kb upstream and downstream). By default, the pipeline orientates the extracted region so that the focal gene is on the positive strand. By default it also omits contigs which have more than a certain number of single nucleotide variants (SNVs) in the focal gene (arbitrary default: 25 SNVs), those containing more than one copy of the focal gene, and those which are are shorter than the requested flanking distance. 

The pipeline then uses \textit{pangraph} \cite{Noll2023} to find homologous stretches of sequence within the flanking regions, which are all analysed together (i.e. homology can be found upstream and downstream simultaneously). \textit{pangraph} infers homology by aligning sequences (with either \textit{minimap2} or \textit{mmseqs2}) and partitioning them into blocks that encapsulate multiple sequence alignments of homologous sequence. (Homology is a binary property: either homologous or non-homologous.) Because \textit{pangraph} was developed with the Mb-scale of whole genomes in mind, it can find homologous stretches within kbs of diverse flanking regions across thousands of sequences on a personal computer in minutes: for a dataset of $n=1,581$ \bla{TEM}-positive assemblies from 24 bacterial genera, building the pangraph of \textasciitilde10kb flanking regions (+/- 5kb) takes less than 3 minutes on a laptop (2 GHz Quad-Core Intel Core i5 processors, 16 GB RAM). \par

Since the aim is to provide a coarse-grained representation of homology for visualisation and quantification, by default the pipeline uses a minimum block size of 100 bp with the \textit{asm10} sensitivity level in \textit{minimap2}. (Since we are primarily concerned with very recent evolutionary events where rearrangements dominate over nucleotide-level variation, we are unconcerned about splitting homologous sequences that have diverged >5\%.) Blocks have a consensus length but can vary slightly in length between input sequences due to smaller-scale differences. Although here we ignore these smaller-scale differences within homologous blocks, precluding the analysis of structural differences below this length scale, a multiple sequence alignment of each block can be reconstructed for more detailed analysis (see \textit{pangraph} \href{https://neherlab.github.io/pangraph/tutorials/tutorial_3/}{documentation}). \par

\subsection*{Visualization}

\noindent The pangraph of the flanking regions in GFA format can be viewed in Bandage, although for most mobile genes the default force-directed layout will produce an uninformative `hairball' due to repeated homology blocks. We therefore produce a linear interactive html representation of the flanking region with blocks coloured by homology (Fig. \ref{fig:workflow}b).  Unique stretches of sequence with no homology within the dataset are coloured grey. Users can highlight a particular block by clicking on it. Optionally, if annotations are provided for the input sequences in gff format these can be projected onto this visualization and toggled on and off (for examples, see the beta-lactamase link).

\subsection*{Quantification}

\noindent The coarse-grained representation of the flanking regions in terms of homologous blocks is our starting point for quantitative analysis. We define two quantities, illustrated schematically in Fig. \ref{fig:workflow}c:

\begin{itemize}
    \item \textit{Breakpoint distance}: for a pairwise comparison of two flanking regions, the distance from the focal gene at which the first non-homologous sequence block occurs. The inverse cumulative distribution function (cdf) of all pairwise distances in a set of sequences can be used to identify breakpoints.
    \item \textit{Block diversity}: at a given distance away from the focal gene, the Shannon entropy of the homologous blocks at that location across the dataset. 
\end{itemize}

\subsection*{Beta-lactamase dataset}

\noindent  We first used prevalence information from the Comprehensive Antibiotic Resistance Database (CARD) v3.1.0 \cite{Alcock2023} to assemble a dataset of complete high-quality chromosomes and plasmids containing at least one gene encoding a beta-lactamase from any of twelve families (CMY, CTX-M, GES, IMP, KPC, NDM, OXA, PER, SHV, TEM, VIM, VEB) according to CARD's `strict' matching criteria. We matched information from these NCBI accessions to get their BioSample ID and also used NCBI entrez to link them to metadata such as species, collection date, host, and geographical location (11 contigs had no associated BioSample). We automatically assigned country names from the NCBI variable \texttt{geo\_loc\_name} and also from \texttt{lat\_lon} where possible. We inspected 42 entries where automatic country names failed and inputted the country manually. Where samples were described as e.g. `USA ex Mexico' we coded this as USA. 

\par 
In the final cleaned metadata: 5,958 contigs (82.5\%) had a collection year, 6,215 (86.0\%) had a country, and 5,764 (79.8\%) had both. Beta-lactamase families can be diverse and contain non-homologous genes {--} notably, the OXA family. Therefore, we used a review of the literature and the classification of clinically important beta-lactamase families \cite{Bush2010} to choose twelve clinically important beta-lactamase genes that have emerged recently as mobile AMR threats as focal genes for our example quantitative analysis (Table \ref{table:bl-genes}). For these twelve focal genes, we included only contigs in our dataset that had sufficient flanking sequence (+/- 5kb) either side of the focal gene with <25 SNVs in the focal gene itself (n=3,362 contigs total). This corresponds to a nucleotide identity cutoff ranging from 96.6\% for \bla{IMP-4} (shortest gene) to 97.8\% for \bla{CMY-2} (longest gene), although in practice nearly all sequences included had <7 SNVs in the focal gene so were >99\% identical at the nucleotide level.
\par

We compiled this dataset in this way before we were aware of \href{https://www.ncbi.nlm.nih.gov/pathogens/microbigge/}{NCBI MicroBIGG-E}. MicroBIGG-E allows users to query genomes already analysed with NCBI AMRFinderPlus as part of the Pathogen Detection Pipeline for a specific gene or genetic element, and then download only its flanking regions across all genomes. This makes it an ideal starting point for flanking region analysis with our pipeline. However, as of 1 August 2023 it only lets the user download flanking regions +/- 2kb. Looking at larger flanking regions currently requires downloading all contigs first.


\section{Application: beta-lactamase genes}

\noindent To demonstrate the scalability of our method, we applied it to twelve different beta-lactamases (Table \ref{table:bl-genes}). Beta-lactam antibiotics are key to modern medicine, accounting for 65\% of prescriptions for injectable antibiotics in the United States \cite{Bush2016}. These antibiotics share a common component: the beta-lactam ring, first seen in the structure of penicillin. Beta-lactamases are a diverse group of enzymes which can break apart the beta-lactam ring by hydrolysis and render beta-lactam antibiotics ineffective. The use of beta-lactam antibiotics therefore exerts a strong selective pressure for sensitive bacteria  to carry beta-lactamases. This effect was first observed immediately after the widespread introduction of penicillin: at one London hospital the proportion of penicillin-resistant \textit{Staphylococcus aureus} carrying the beta-lactamase \textit{bla}Z increased from 14\% in 1946 to 38\% the following year \cite{Barber1947}. The increase in the prevalence of beta-lactamases and their adaptation to hydrolyse successive generations of beta-lactams to  is one of the clearest real-world examples of rapid evolution.\par

Beta-lactamases are a key clinical problem in Gram-negative species \cite{Bush2020} and recent estimates suggest much higher prevalences of beta-lactamases in the Global South \cite{Global2022}. WHO has designated Gram-negative species priority pathogens for the development of new antibiotics \cite{WHO2017}. In recent decades, many newly-described beta-lactamases have been identified in clinical bacteria \cite{Bush2020}. A particular concern are emerging extended-spectrum beta-lactamases (ESBLs) which confer resistance to a range of beta-lactams \cite{Livermore2008} and commonly spread on MGEs. In some cases these genes can be identified as having been mobilized by MGEs from the chromosomes of environmental bacteria {--} \textcite{Partridge2011a} gives a list of the probable ancestral species for many beta-lactamases. Many mobile beta-lactamases are speculated to have originated from a single mobilization event. Mobile beta-lactamases therefore provide repeated examples of a common pattern: mobilization of a chromosomal gene, followed by diversification of its flanking regions as it spreads through bacterial pangenomes into new genomic contexts under strong selective pressure.\par

We selected twelve clinically important beta-lactamases as focal genes, then downloaded complete high-quality chromosomes and plasmids to run through our pipeline looking at +/- 5kb flanking regions (see Methods). A single amino acid change in a beta-lactamase is enough to denote a new numbered variant (e.g.  \bla{NDM-1} and \bla{NDM-2}) although identical amino acids can still have synonymous SNVs in their nucleotide sequence. We arbitarily chose 25 SNVs in the focal gene (i.e. nucleotide) as a cutoff to include highly-related variants, approximately corresponding to a 97.5\% nucleotide identity cutoff (in practice nearly all included had <7 SNVs so >99\%). \par

The interactive visualizations produced are available at \href{https://liampshaw.github.io/flanking-regions}{https://liampshaw.github.io/flanking-regions}. For the remainder of this paper, we explore three general quantitative findings that hold {--} more or less {--} across these diverse genes, and comment on the possible processes that generate structural diversity.

\subsubsection{The accumulation of nucleotide-level variation correlates with the breakdown of flanking homology}

\begin{figure}
    \centering
    \includegraphics[width=0.8\linewidth]{figs/FINAL-ctx-m-65-distances-figure.pdf}
    % output/pangraph/CTX-M-65/plots/CARD_plot_breakpoint_distances-compare-to-focal-gene.pdf 
    \caption{\textbf{The breakdown of homology in the flanking regions of \bla{CTX-M-65} and closely-related genes.} (a) Visualization of homology in the 5kb flanking regions of \bla{CTX-M-65} and two closely-related genes that encode different CTX-M protein variants, \bla{CTX-M-24} (1 SNV apart) and \bla{CTX-M-14} (2 SNVs apart). Only a subset of 15 sequences are shown for each gene. Grey blocks have no homology within the wider dataset. (b) The inverse cumulative distribution function of pairwise comparisons of distance to the first breakpoint (first occurrence of non-homologous sequence) between $n=434$ sequences carrying any \bla{CTX-M} gene with <25 SNVs to \bla{CTX-M-65}, stratified by the number of SNVs in the focal gene relative to \bla{CTX-M-65} (only comparisons involving \bla{CTX-M-65} are shown). The same pattern is seen when picking a single isolate for each year/country/genus combination to control for potential sampling bias (Fig. S1).}
    \label{fig:CTX-M-65}
\end{figure}


\begin{figure*}
    \centering
    \includegraphics[width=0.8\linewidth]{figs/FINAL-figure-AUC-top-bl-final.pdf}
    \caption{\textbf{Structural diversity correlates with nucleotide-level variation.} Across different beta-lactamases the presence of any SNV in the focal gene is generally correlated with a shorter flanking region overlap (sum of average upstream and downstream pairwise breakpoint distances). (a) shows linear fits (\texttt{stat\_smooth}) and (b) average per SNV-level comparison for $n=7$ genes with >100 genomes in the initial dataset (Table \ref{table:bl-genes}). Pairwise comparisons have been deduplicated for year/country/genus combinations before plotting. The flanking region overlap is the sum of the upstream and downstream average breakpoint distances. Number of data points for each gene (after deduplicating): \bla{CMY-2} (n=108), \bla{CTX-M-15} (n=274), \bla{CTX-M-65} (n=119), \bla{GES-1} (n=15), \bla{KPC-2} (n=161), \bla{NDM-1} (n=175), \bla{OXA-48} (n=66), \bla{TEM-1} (n=514).}
    \label{fig:AUC-decay}
\end{figure*}

\noindent In a simple model of a gene that is mobilized from a chromosomal background and then spreads horizontally on a MGE, the accumulation of single nucleotide variants (SNVs) in the gene itself should be a `slow' molecular clock in contrast to the `fast' rearrangements that happen in its flanking regions. The two would be expected to be broadly correlated: more SNVs in the gene suggest that more time has elapsed, and so greater structural diversity should have accumulated in its flanking regions.\par

The CTX-M-9-like beta-lactamases provide a clear example of this pattern. Arbitrarily choosing the CTX-M-9-like gene \bla{CTX-M-65} as a focal gene and searching for sequences with variants, we find a strong correlation between SNVs in the gene and breakpoint distances in flanking regions (Fig. \ref{fig:CTX-M-65}). This does not appear to be due to repeated sequencing due to outbreaks, because the pattern persists if we include one random isolate from each year/country/genus combination (Fig. S1). There is still some shared sequence around almost all CTX-M-9-like genes, supporting their common origin in some previous mobilisation. Indeed, previously \textcite{Olson2005} found a chromosomal beta-lactamase in \textit{Kluyvera georgiana} that shared 100\% amino acid identity with CTX-M-14 and was in a 2.7kb region with 99\% nucleotide identity to the complex class 1 integron In60, arguing therefore that \textit{K. georgiana} was the likely source for the progenitor of the CTX-M-9-like group through mobilisation. However, since then different CTX-M-9-like beta-lactamases have diverged in their mobilisation.\footnote{This suggests that \bla{CTX-M-14} is the ancestral gene, so would be a better choice of focal gene than \bla{CTX-M-65} as we arbitrarily chose here. In fact CTX-M-14 has been suggested to have evolved twice by convergent evolution with different flanking regions for the two nucleotide variants \cite{Navarro2007}.}\par

With a model of a single initial mobilization, all subsequent disruptions of shared sequence are due to subsequent insertions (or deletions) which introduce non-homologous sequence. There are two `modes' of this breakdown of shared sequence in the cumulative distribution of breakpoint distances: gradual decay and sharp breakpoints (Fig. \ref{fig:CTX-M-65}). A sharp breakpoint across pairwise comparisons might be suggestive of a consistent `common block', which could be a coherent MGE inserting into multiple genomic backgrounds or a gene cassette in an integron's gene array; gradual decay is suggestive of a `fossilized' MGE that is undergoing degradation via the introduction of non-homologous sequence in its flanking regions at random points. These patterns may be driven by the same underlying processes.\par

The area under the curve (AUC) of the distribution of breakpoint distances for shared sequence between pairs of sequences is equal to the average breakpoint distance length. We can therefore use this to quantify the rough correlation between nucleotide-level variation and structural diversity. By stratifying pairwise comparisons between sequences based on the SNVs in the focal gene, there is a tendency for greater diversification around the focal gene with more SNVs (Fig. \ref{fig:AUC-decay}). 

\begin{figure*}
    \centering
    \includegraphics[width=0.8\linewidth]{figs/figure-NDM-curves.pdf}
    % output/pangraph/CTX-M-65/plots/CARD_plot_breakpoint_distances-compare-to-focal-gene.pdf 
    \caption{\textbf{Homology decay, block diversity, and transposase density around \bla{NDM-1}.} Homology decay in terms of pairwise distance to first breakpoint (black), block diversity (green) and transposase density from annotations as a proxy for IS density (blue). All curves are normalized (block diversity to log(N), transposase density to the point of maximum density in the flanking region shown (290/333 sequences). Left shows upstream, right shows downstream.}
    \label{fig:NDM-1-curves}
\end{figure*}

\subsubsection{Linking structural diversity to annotations}

\noindent It is well-established that insertion sequences (ISs), which use transposases to catalyze DNA cleavage and strand transfer leading to movements, are fundamental to the mobilization of AMR genes and contribute to the complexity of their flanking regions. Repeats in ISs are often the reason why flanking regions cannot be assembled in genome assemblies. We used existing annotations from NCBI to identify the locations of transposases (as a proxy for IS presence). \par
Plotting the block diversity and the breakpoint distance distribution with the transposase density revealed interesting patterns (Fig. S2). As an example: for \bla{NDM-1}, the highest density of transposases is upstream, with an associated immediate breakdown of shared sequence (Fig. \ref{fig:NDM-1-curves}). After this breakdown of homology, there is not much further diversification, a signature that the gene has become stabilized in different backgrounds. Downstream, the breakdown of the shared ancestral background is more gradual. This recapitulates the more detailed analysis of \textcite{Acman2022}.\par


\begin{figure}
    \centering
    \includegraphics[width=0.95\linewidth]{figs/FINAL-figure-AUC-upstream-downstream.pdf}
    % full dataset: figs/figure-breakpoint-AUCs-all.pdf
    \caption{\textbf{Comparing upstream and downstream breakpoint distances.} Results show average downstream and upstream breakpoint distances for each focal gene, including only comparisons between identical focal genes (0 SNVs) after metadata deduplication for year/genus/country. There seems to be a bias for greater conservation downstream. There is no overall significant difference at $\alpha=0.05$ (7/12 downstream>upstream; Wilcoxon signed-rank exact test $V=57$, $p=0.18$). However, when integron-associated genes (grey) are removed, the difference becomes stronger (6/8 downstream>upstream, $V=32$, $p=0.055$.}
    \label{fig:auc-correlation-up-down}
\end{figure}

\subsubsection{A bias for greater structural diversity upstream}
\noindent Using the distribution of breakpoint distances, we found an asymmetry between upstream and downstream flanking regions: after removing integron-associated genes, there is a lower average breakpoint distance upstream, indicative of greater structural diversity (Fig. \ref{fig:auc-correlation-up-down}). This suggests that, considering mobile beta-lactamases associated with ISs, there is a bias in the realized rate of introduction of new structural diversity in flanking regions: it has a tendency to be higher upstream of a mobile gene. Below we speculate as to the cause of this asymmetry.

\section{Discussion}

\noindent We have aimed to provide a starting point to quantitatively investigate structural diversity around mobile genes. By using \textit{pangraph} to find homologous sequences in flanking regions without requiring annotation information, we have constructed a pipeline that quickly produces interactive visualizations that can be explored by researchers to make sense of these complex regions, as well as computing some basic summary statistics. To demonstrate our pipeline, we applied it to the flanking regions of twelve different beta-lactamase genes in public high-quality assemblies. Our observations recapitulate previous knowledge about individual beta-lactamase genes but at scale, and provide evidence of general patterns.

We wish to highlight three limitations of our approach. First, we stress that while this approach gives a way to quantify structural diversity around a gene, it is not intended to infer how a specific instance of the gene is \textit{currently} mobilized. For simplicity we have analyzed 5kb flanking regions, but these are unlikely to capture all information on mobilization: in Gram-negative species mobilized genes tend to cluster in `multiresistance regions’ which can be tens of kilobases in size \cite{Partridge2021}. Second, our approach is coarse-grained {--} in the sense that \textit{pangraph} has a minimum size limit for the size of homologous blocks. This means that traces of evolution at smaller scales will be missed by considering homologous blocks as ‘identical’, for example transposition events which are associated with small-scale (2-bp) insertions. More fine-grained analysis is possible using the multiple-sequence alignments that are generated for each block by \textit{pangraph}, although a bespoke analysis (e.g. an alignment against a known structure to identify small-scale changes) would always be expected to be superior. In particular, an issue in \textit{pangraph} is that breaks in homology involving small blocks close to the minimum size (100bp) appear more unreliable, and we are investigating how best to process these. Third, our analysis uses public genomes which are a biased sample, so all inferences about evolutionary events should be viewed with appropriate scepticism.

We are far from being able to build a quantitative evolutionary model of the structural diversity around mobile genes. However, the patterns we find for beta-lactamases suggest some general principles about the evolution of flanking regions, which appear consistent with what has previously been noted in the literature.\par 

First, nucleotide-level variation in the mobile gene itself does indeed appear to be a `slow' molecular clock compared to the much faster rates of rearrangements in flanking regions. Although it should be noted that this variation is itself under selection and is not a neutral clock, since single SNVs can change the hydrolytic profile of beta-lactamases, this suggests a quantitative connection between mutational and non-mutational evolution that could be fruitful for further exploration. Second, we confirm that ISs can be linked to high levels of structural diversity around mobile genes, likely driven by homologous recombination as well as transposition. Third, we found an asymmetry with direction, with greater conservation of downstream flanking regions in 8/12 beta-lactamases, meaning greater structural diversity upstream. Speculatively, we propose that this observation could be a signature of the mobilization dynamics of IS-associated beta-lactamases.\par 

It is known that many beta-lactamases have been mobilized from chromosomal backgrounds and are weakly expressed in their native genomic context. However, when selective pressure for antibiotics requires higher expression, the mobilization of these genes is linked to enhanced expression. It was noted in the 1990s that the inverted repeats around ISs often contain promoters, and that ISs are often found immediately upstream of beta-lactamase genes e.g. for \bla{TEM-6} \cite{Goussard1991} or \bla{CTX-M-14} \cite{Cao2002}. Similarly, Poirel et al. observed that the upstream insertion sequence ISEcp1 which mobilizes \bla{CTX-M-19} contains a strong promoter, simultaneously mobilising the resistance gene and providing strong expression \cite{Poirel2003}. The spread of beta-lactamases is due to strong selection for increased expression: ISs can provide this increased expression. In this model, since a mobile beta-lactamase depends initially on the upstream insertion of at least one IS, that location will then be a hotspot for the creation of new structural diversity, whether through subsequent insertion to the same target site or homologous recombination between common regions of ISs. Where genes were associated not with an IS but an integron (which is the case for \bla{IMP-4}, \bla{OXA-10}, \bla{VIM-1}, \bla{GES-1}), we did not observe this asymmetry. More widely, AMR genes tend to accumulate into multi-resistance regions, which has been suggested to be driven by homologous recombination between common components such as transposases \cite{Partridge2011b}. Repeated rounds of homologous recombination and insertion that degrade `active' mobile MGEs may play a similar role in the creation of multi-resistance regions to that suggested to be responsible for the creation of defence islands in chromosomes \cite{Rocha2022}.  

\subsection*{Summary}

\noindent The approach we outline for visualizing and quantifying structural diversity in flanking regions is applicable to any gene and scales to thousands of sequences on the kb-scale. We have applied it to recently emerged beta-lactamases as an example of mobile genes. Each mobile gene is worth detailed study on its own and homology visualizations can help understand the patterns in its flanking regions. Large-scale analysis across genes can reveal patterns consistent with similar underlying processes, suggesting conclusions consistent with the existing literature. Future quantitative methods will allow us to better understand the dynamics that govern the arrival and establishment of genes within pangenomes. 

\section*{Funding}

\noindent Liam Shaw is a Sir Henry Wellcome Postdoctoral Fellow funded by Wellcome (Grant 220422/Z/20/Z). Richard Neher is funded by the University of Basel and the SNSF. 

\section*{Acknowledgements}

\noindent The authors thank Sally Partridge, Zamin Iqbal and Will Matlock for helpful feedback on a draft version, and Marco Molari for ongoing and stimulating discussions. Liam Shaw thanks John Lees and Craig MacLean for feedback on visualizations. 

\newpage
\section*{Software and data availability}

\noindent \begin{itemize}
\item Analysis pipeline (release: v0.2): \\\href{https://github.com/liampshaw/mobile-gene-regions}{https://github.com/liampshaw/mobile-gene-regions}
\item Interactive plots for n=12 beta-lactamases:\\\href{https://liampshaw.github.io/flanking-regions}{https://liampshaw.github.io/flanking-regions}
\item Beta-lactamase dataset:\\\href{https://doi.org/10.5281/zenodo.8208376}{https://doi.org/10.5281/zenodo.8208376}
\end{itemize}

\if{
\subsection*{Accuracy of block-based breakpoint distance}

Pangraph uses a greedy algorithm to efficienctly cluster blocks of homologous sequence across a diverse dataset. In a small number of cases we have observed that this can lead to inaccuracies in the resulting homology blocks, including a shared stretch of sequence in two otherwise diverse genomes being split into different homology blocks (Fig. \textit{example}). We validated using the homology blocks to calculate breakpoint distances compared to using direct pairwise comparisons with \textit{minimap2} (options: `-r 10 -X --no-long-join`) and found that the overall distribution was similar. We recommend that for specific comparisons of two sequences, a detailed local alignment is always performed. 
}\fi



\newpage
\bibliography{cite}

\end{document}